activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb

```diff
@@ -9,9 +9,9 @@ module ConnectionAdapters #:nodoc:
     # are typically created by methods in TableDefinition, and added to the
     # +columns+ attribute of said TableDefinition object, in order to be used
     # for generating a number of table creation or table changing SQL statements.
-    ColumnDefinition = Struct.new(:name, :type, :limit, :precision, :scale, :default, :null, :first, :after, :auto_increment, :primary_key, :collation, :sql_type, :comment, :as) do # :nodoc:
-      def primary_key?
-        primary_key || type.to_sym == :primary_key
+    ColumnDefinition = Struct.new(:name, :type, :options, :sql_type) do # :nodoc:
+      def [](key)
+        options[key]
```

Comment:
This poses a compatibility challenge for third party database adapters. Could we offer an gentler upgrade path? Possibly by deprecating the old methods:
```ruby
class ColumnDefinition
  attr_reader :name, :type, :sql_type, :options

  def initialize(*args, options: nil)
    if options
      initialize_with_options(*args, options)
    else
      initialize_with_struct_args(*args)
    end
  end

  def initialize_with_options(name, type, sql_type, options)
    @name, @type, @options, @sql_type = name, type, options, sql_type
  end

  def initialize_with_struct_args(name, type, limit, etc…)
    ActiveSupport::Deprecation.warn …
    @name, @type, @sql_type = name, type, sql_type
    @options = { limit: limit, precision: precision, … }
  end

  %i[ limit precision … ].each do |option_name|
    define_method(option_name) do
      ActiveSupport::Deprecation.warn …
      options[option_name]
    end

    define_method("#{option_name}=") do |value|
      ActiveSupport::Deprecation.warn …
      options[option_name] = value
    end
  end
end
```